source/connection_h1.c

```diff
@@ -1130,63 +1128,65 @@ static int s_handler_process_read_message(
     struct aws_io_message *message) {
 
     struct h1_connection *connection = handler->impl;
-
-    if (connection->thread_data.is_shutting_down) {
-        AWS_LOGF_ERROR(
-            AWS_LS_HTTP_CONNECTION,
-            "id=%p: Cannot process message because connection is shutting down.",
-            (void *)&connection->base);
-
-        aws_raise_error(AWS_ERROR_HTTP_CONNECTION_CLOSED);
-        goto error;
-    }
-
-    if (!connection->thread_data.incoming_stream) {
-        /* TODO: Server connection would create new request-handler stream at this point. */
-
-        AWS_LOGF_ERROR(
-            AWS_LS_HTTP_CONNECTION,
-            "id=%p: Cannot process message because no requests are currently awaiting response, closing connection.",
-            (void *)&connection->base);
-
-        aws_raise_error(AWS_ERROR_HTTP_INVALID_PARSE_STATE);
-        goto error;
-    }
+    int err;
 
     /* By default, we will increment the read window by the same amount we just read in.
      * However, users have the opportunity to tweak this number in their aws_http_on_incoming_body_fn() callback. */
     connection->thread_data.incoming_message_window_update = message->message_data.len;
 
-    /* Decoder will invoke the internal s_decoder_X callbacks, which in turn invoke user callbacks */
     AWS_LOGF_TRACE(
         AWS_LS_HTTP_CONNECTION,
         "id=%p: Begin processing message of size %zu.",
         (void *)&connection->base,
         message->message_data.len);
 
-    size_t decoded_len = 0;
-    int err = aws_http_decode(
-        connection->thread_data.incoming_stream_decoder,
-        message->message_data.buffer,
-        message->message_data.len,
-        &decoded_len);
-    if (err) {
-        AWS_LOGF_ERROR(
-            AWS_LS_HTTP_CONNECTION,
-            "id=%p: Message processing failed, error %d (%s). Closing connection.",
-            (void *)&connection->base,
-            err,
-            aws_error_name(err));
+    /* Run decoder until all message data is processed */
+    struct aws_byte_cursor message_cursor = aws_byte_cursor_from_buf(&message->message_data);
+    while (message_cursor.len > 0) {
```

Comment:
How much data could conceivably be backed up here? This appears to have removed the shutdown check, so that would mean shutdown has to decode potentially a lot of data and deliver it before completing? Is that desirable?

Reply:
We're actually checking for shutdown more frequently now.

It used to be that a call to decode() would process ALL the data passed in, possibly processing multiple responses.

Now decode() exits after each response is processed, so we call it in a loop, and in each loop we check that the connection isn't shutting down, and that there is in fact a pending-request to receive the response